fuse/server.go

```diff
@@ -28,6 +28,21 @@ type Server struct {
 	// writeMu serializes close and notify writes
 	writeMu sync.Mutex
 
+	// forgetMu prevents concurrent execution of FORGET, BATCH_FORGET
+	// and LOOKUP. The problem is:
+	//
+	// Thread 1            Thread 2
+	// ----------------    -------------------
+	// BATCH_FORGET 1,2
+	// --> forget 1
+	//                     LOOKUP "foo"
+	//                     --> return NodeId 2
+	// --> forget 2
+	//
+	// At this point the kernel holds a reference to NodeId 2 that has
+	// been forgotten inside go-fuse.
+	forgetMu sync.RWMutex
+
 	// I/O with kernel and daemon.
 	mountFd int
 
```

Comment:
do you have any proof that this is a problem?

forget and lookup have reference counts. In your scenario, the count for  "foo" starts at 1, goes to 2 (lookup "foo") and then back to 1 (forget). 

The kernel will have forgotten node 2, but will know about it once the Lookup result is processed.